### PR TITLE
Truncate pubsub names if greater than 255 chars

### LINKED ISSUE #831: #1 Broker reconciler doesn't validate length of pubsub resource names
**Describe the bug**
Pubsub resource names can't be longer than 255 characters. The resource names generated for brokers and triggers can be longer than that if their K8s names are especially long.

Similar to kmeta.ChildName, we need a strategy to truncate those names safely if they are greater than the max length. We probably shouldn't use kmeta.ChildName directly because its max length is too short.

 **Expected behavior**
Generated names are never longer than 255 characters.

 **To Reproduce**
Namespace max length: 63
Broker and trigger name max length: 253
UID length: 36
Prefix and separators: 10

 **Knative-GCP release version**
HEAD

 **Additional context**

### PR DESCRIPTION
**Pull request by grantr**
Fixes #831
Supersedes #942

Since we append the UID to the resource name, there's no need to use a hashed truncate to guarantee uniqueness of the truncated name as in `kmeta.ChildName`.

## Proposed Changes

- Truncate pubsub resource names deterministically if they would be greater than 255 characters

### CODE REVIEW

File: pkg/reconciler/broker/resources/names_test.go [hunk 1]
```diff
@@ -46,12 +49,25 @@ func TestGenerateDecouplingTopicName(t *testing.T) {
 		n:    "more-dashes",
 		uid:  testUID,
 		want: fmt.Sprintf("cre-bkr_with-dashes_more-dashes_%s", testUID),
+	}, {
+		ns:   maxNamespace,
+		n:    maxName,
+		uid:  testUID,
+		want: fmt.Sprintf("cre-bkr_%s_%s_%s", maxNamespace, strings.Repeat("n", 146), testUID),
```
Comment (liu-cong) [inline pkg/reconciler/broker/resources/names_test.go:56]: nit: define const maxNameLength=146 for better readablility.

**Comment by grantr** [comment]
/hold cancel

@liu-cong RFAL

**Pull request merged**